File: src/main/java/seremis/geninfusion/network/packet/PacketEntityData.java
package seremis.geninfusion.network.packet;

import seremis.geninfusion.entity.GIEntity;
import seremis.geninfusion.entity.GIEntityLiving;
import io.netty.buffer.ByteBuf;
import io.netty.channel.ChannelHandlerContext;
import net.minecraft.entity.Entity;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.network.PacketBuffer;

public class PacketEntityData extends AbstractPacket {

    int entityId;
    byte[] data;
    int length;
    int id;
	
	public PacketEntityData() {
		
	}
	
	public PacketEntityData(byte[] data, int id, int entityId) {
        this.entityId = entityId;
        this.data = data;
        this.id = id;

        length = data.length;
    }

	@Override
	public void encodeInto(ChannelHandlerContext ctx, ByteBuf buffer) {
        PacketBuffer pbuff = new PacketBuffer(buffer);
        pbuff.writeByte(id);
        pbuff.writeInt(length);
        pbuff.writeBytes(data);
	}

	@Override
	public void decodeInto(ChannelHandlerContext ctx, ByteBuf buffer) {
		PacketBuffer pbuff = new PacketBuffer(buffer);
        id = pbuff.readByte();
        length = pbuff.readInt();
        data = new byte[length];
        pbuff.readBytes(data);		
	}

	@Override
	public void handleClientSide(EntityPlayer player) {
		Entity ent = player.worldObj.getEntityByID(entityId);
		
		if(ent != null && ent instanceof GIEntity) {
			((GIEntity) ent).receivePacketOnClient(id, data);
		} else if(ent != null && ent instanceof GIEntityLiving) {
			((GIEntityLiving) ent).receivePacketOnClient(id, data);
		}
	}

	@Override
	public void handleServerSide(EntityPlayer player) {
		Entity ent = player.worldObj.getEntityByID(entityId);
		
		if(ent != null && ent instanceof GIEntity) {
			((GIEntity) ent).receivePacketOnServer(id, data);
		} else if(ent != null && ent instanceof GIEntityLiving) {
			((GIEntityLiving) ent).receivePacketOnServer(id, data);
		}
	}
}

File: src/main/java/seremis/geninfusion/network/packet/PacketEntityData.scala
package seremis.geninfusion.network.packet

import cpw.mods.fml.common.network.simpleimpl.{IMessage, IMessageHandler, MessageContext}
import cpw.mods.fml.relauncher.Side
import io.netty.buffer.ByteBuf
import net.minecraft.client.Minecraft
import seremis.geninfusion.entity.GIEntity

class PacketEntityData(var data: Array[Byte], var id: Int, var entityId: Int) extends IMessage {

    var length: Int = data.length

    override def fromBytes(buf: ByteBuf) {
        entityId = buf.readInt()
        id = buf.readByte()
        length = buf.readInt()
        data = Array.ofDim[Byte](length)
        buf.readBytes(data)
    }

    override def toBytes(buf: ByteBuf) {
        buf.writeInt(entityId)
        buf.writeByte(id)
        buf.writeInt(length)
        buf.writeBytes(data)
    }
}

object PacketEntityData {

    class Handler extends IMessageHandler[PacketEntityData, IMessage] {

        override def onMessage(message: PacketEntityData, ctx: MessageContext): IMessage = {
            if (ctx.side == Side.CLIENT) {
                val ent = Minecraft.getMinecraft.theWorld.getEntityByID(message.entityId)

                if (ent != null && ent.isInstanceOf[GIEntity]) {
                    ent.asInstanceOf[GIEntity].receivePacketOnServer(message.id, message.data)
                }
            } else {
                val ent = ctx.getServerHandler.playerEntity.worldObj.getEntityByID(message.entityId)

                if (ent != null && ent.isInstanceOf[GIEntity]) {
                    ent.asInstanceOf[GIEntity].receivePacketOnServer(message.id, message.data)
                }
            }
            null
        }
    }
}